0020-bluetooth-payment-protocol/0020-bluetooth-payment-protocol.md

```diff
@@ -0,0 +1,117 @@
+  - __unauthenticated__: Payment was canceled because of invalid password
+  - __canceled__: Payment was canceled by the Customer. 
+  - __complete__: Payment complete. Read `:iprAuthorization` for the signed packet
+- `:action` (Write)
+  - __sign__: Initiate signing after `:packet`, `:packetSignature`, `:merchantPublicKey` and `:merchantCertificate` have been written.
+- `:ipr`: (Write Long Characteristic)
+  - ILP Packet containing destination Address and Cost
+- `:iprSignature`: (Write Long Characteristic)
+  - Merchant signature on IPR.
+- `:merchantPublicKey`: (Write Long Characteristic)
+- `:merchantCertificate`: (Write Long Characteristic)
+- `:iprAuthorization`: (Read Long Characteristic)
+  - Device signature on IPR
+
+## Merchant Setup
+
+1. Merchant generates certificate with Ed25519 public key
+2. Merchant submits certificate, ILP account address, and per-payment limit to a Merchant Certificate Authority
+3. Certificate Authority signs merchant details
+
+## Customer Device Setup
+
+1. Device manufacturer embeds unique Ed25519 public / private key pair in device. The manufacturer MUST ensure the device is tamper-proof.
+2. Device manufacturer signs the device public key with their own key pair.
+3. Customer submits the device public key and manufacturer's signature to their ledger operator to authorize the device to spend funds from their account.
+4. If the device manufacturer is trusted to produce secure devices, the ledger operator signs the device public key and a per-payment spending limit with their own key.
+5. Customer stores the ledger operator's signature, certificate, per-payment and total spending limit on the device (**TODO**: would it make more sense to have the manufacturer include the ledger signature and cert in the device?).
+
+## Payment Flow
+
+**[MERC]** indicates steps taken by the Merchant
+
+**[CUST]** indicates steps taken by the Customer's device
+
+1. **[MERC]** Listen to Broadcasts of the [ILP Account Information](#ILPAccountInformation). List customers.
+2. **[MERC]** Select customer for payment, and initiate a connection over BLE.
+3. **[MERC]** Generate IPR with __account__ and __amount__.
+4. **[MERC]** Sign the IPR using their Ed25519 keypair
+5. **[MERC]** Write to:
+  1. `:ipr` with IPR
+  2. `:iprSignature` with Signature of IPR
+  3. `:merchantPublicKey` with the Merchant Public Key 
+  4. `:merchantCertificate` with the Certificate signed by a Certificate Authority
+6. **[MERC]** Write __sign__ to `:action`, to indicate data is ready to be validated/signed.
+7. **[MERC]** Listen to notifications on `:status`
+8. **[CUST]**  Write __"authenticating"__ to `:status`
+9. **[CUST]** Validates payload on Characteristics. If:
+  1. [CA is NOT trusted:](#) write __"invalid"__ to `:status`. Close connection
+  2. [Payment NOT within limits:](#) write __"invalid"__ to `:status`. Close connection
```

Comment:
That would require information about the real value of the product, such as "US$4,00". IPR protocol would need to have information about source or destination amount, with currency and scale so that it can be fulfilled later on